File: SMART/app/src/main/java/com/cs565project/smart/db/AppDao.java
package com.cs565project.smart.db;

import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Delete;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Update;

import com.cs565project.smart.db.entities.AppDetails;
import com.cs565project.smart.db.entities.Category;
import com.cs565project.smart.db.entities.DailyAppUsage;
import com.cs565project.smart.db.entities.MoodLog;
import com.cs565project.smart.db.entities.RecommendationActivity;

import java.util.Date;
import java.util.List;

/**
 * Data access object which provides queries, inserts, updates and deletes in the app database.
 */
@Dao
public interface AppDao {
    @Query("SELECT * FROM DailyAppUsage WHERE packageName = (:packageName) AND date = (:date) LIMIT 1")
    DailyAppUsage getAppUsage(String packageName, Date date);

    @Query("SELECT * FROM DailyAppUsage WHERE date = (:date)")
    List<DailyAppUsage> getAppUsage(Date date);

    @Query("SELECT * FROM DailyAppUsage WHERE date >= (:startDate) AND date <= (:endDate)")
    List<DailyAppUsage> getAppUsage(Date startDate, Date endDate);

    @Query("SELECT * FROM DailyAppUsage WHERE packageName = (:packageName)")
    List<DailyAppUsage> getAppUsage(String packageName);

    @Query("SELECT MIN(date) FROM DailyAppUsage")
    Date getUsageDataStartDate();

    @Query("SELECT * FROM AppDetails")
    List<AppDetails> getAppDetails();

    @Query("SELECT packageName FROM AppDetails")
    List<String> getAppPackageNames();

    @Query("SELECT * FROM AppDetails WHERE packageName IN (:packageNames)")
    List<AppDetails> getAppDetails(List<String> packageNames);

    @Query("SELECT * FROM AppDetails WHERE packageName = (:packageName) LIMIT 1")
    AppDetails getAppDetails(String packageName);

    @Query("SELECT * FROM AppDetails WHERE thresholdTime > -1")
    List<AppDetails> getRestrictedAppDetails();

    @Query("SELECT * FROM MoodLog WHERE dateTime = (:date)")
    MoodLog getMoodLog(Date date);

    @Query("SELECT * FROM MoodLog WHERE dateTime >= (:startDate) AND dateTime <= (:endDate)")
    List<MoodLog> getMoodLog(Date startDate, Date endDate);

    @Query("SELECT * FROM MoodLog")
    List<MoodLog> getAllMoodLog();

    @Query("SELECT * FROM RecommendationActivity")
    List<RecommendationActivity> getRecommendationActivities();

    @Query("SELECT * FROM RecommendationActivity WHERE activityType = (:activityType)")
    List<RecommendationActivity> getRecommendationActivities(String activityType);

    @Query("SELECT * FROM Category")
    List<Category> getCategories();

    @Query("SELECT name FROM Category WHERE shouldRestrict = (:shouldRestrict)")
    List<String> getCategories(boolean shouldRestrict);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertAppUsage(DailyAppUsage... appUsages);
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertAppUsage(List<DailyAppUsage> appUsages);
    @Insert
    void insertAppDetails(AppDetails... apps);
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertAppDetails(List<AppDetails> apps);
    @Insert
    void insertMood(MoodLog... mood);
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertRecommendationActivity(RecommendationActivity activity);
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertCategory(Category category);

    @Update
    void updateAppDetails(AppDetails... apps);
    @Update
    void updateAppUsage(DailyAppUsage... appUsages);
    @Update
    void updateRecommendationActivity(RecommendationActivity activity);

    @Delete
    void deleteAppUsage(DailyAppUsage... appUsages);
    @Delete
    void deleteAppDetails(AppDetails... apps);
}

File: SMART/app/src/main/java/com/cs565project/smart/db/AppDao.kt
package com.cs565project.smart.db

import android.arch.lifecycle.LiveData
import android.arch.persistence.room.*
import com.cs565project.smart.db.entities.*
import java.util.*

/**
 * Data access object which provides queries, inserts, updates and deletes in the app database.
 */
@Dao
interface AppDao {

    @get:Query("SELECT MIN(date) FROM DailyAppUsage")
    val usageDataStartDate: Date

    @get:Query("SELECT * FROM AppDetails")
    val appDetails: List<AppDetails>

    @get:Query("SELECT packageName FROM AppDetails")
    val appPackageNames: List<String>

    @get:Query("SELECT * FROM AppDetails WHERE thresholdTime > -1")
    val restrictedAppDetails: List<AppDetails>

    @get:Query("SELECT * FROM MoodLog")
    val allMoodLog: List<MoodLog>

    @get:Query("SELECT * FROM RecommendationActivity")
    val recommendationActivities: List<RecommendationActivity>

    @get:Query("SELECT * FROM Category")
    val categories: List<Category>

    @Query("SELECT * FROM DailyAppUsage WHERE packageName = (:packageName) AND date = (:date) LIMIT 1")
    fun getAppUsage(packageName: String, date: Date): DailyAppUsage

    @Query("SELECT * FROM DailyAppUsage WHERE date = (:date)")
    fun getAppUsage(date: Date): List<DailyAppUsage>

    @Query("SELECT * FROM DailyAppUsage WHERE date = (:date)")
    fun getAppUsageLiveData(date: Date): LiveData<List<DailyAppUsage>>

    @Query("SELECT * FROM DailyAppUsage WHERE date >= (:startDate) AND date <= (:endDate)")
    fun getAppUsage(startDate: Date, endDate: Date): List<DailyAppUsage>

    @Query("SELECT * FROM DailyAppUsage WHERE packageName = (:packageName)")
    fun getAppUsage(packageName: String): List<DailyAppUsage>

    @Query("SELECT * FROM AppDetails WHERE packageName IN (:packageNames)")
    fun getAppDetails(packageNames: List<String>): List<AppDetails>

    @Query("SELECT * FROM AppDetails WHERE packageName = (:packageName) LIMIT 1")
    fun getAppDetails(packageName: String): AppDetails

    @Query("SELECT * FROM AppDetails WHERE packageName IN (:packageNames)")
    fun getAppDetailsLiveData(packageNames: List<String>): LiveData<List<AppDetails>>

    @Query("SELECT * FROM MoodLog WHERE date = (:date)")
    fun getMoodLog(date: Date): MoodLog

    @Query("SELECT * FROM MoodLog WHERE date >= (:startDate) AND date <= (:endDate)")
    fun getMoodLog(startDate: Date, endDate: Date): List<MoodLog>

    @Query("SELECT * FROM RecommendationActivity WHERE activityType = (:activityType)")
    fun getRecommendationActivities(activityType: String): List<RecommendationActivity>

    @Query("SELECT name FROM Category WHERE shouldRestrict = (:shouldRestrict)")
    fun getCategories(shouldRestrict: Boolean): List<String>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertAppUsage(vararg appUsages: DailyAppUsage)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertAppUsage(appUsages: List<DailyAppUsage>)

    @Insert
    fun insertAppDetails(vararg apps: AppDetails)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertAppDetails(apps: List<AppDetails>)

    @Insert
    fun insertMood(vararg mood: MoodLog)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertRecommendationActivity(activity: RecommendationActivity)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertCategory(category: Category)

    @Update
    fun updateAppDetails(vararg apps: AppDetails)

    @Update
    fun updateAppUsage(vararg appUsages: DailyAppUsage)

    @Update
    fun updateRecommendationActivity(activity: RecommendationActivity)

    @Delete
    fun deleteAppUsage(vararg appUsages: DailyAppUsage)

    @Delete
    fun deleteAppDetails(vararg apps: AppDetails)

    @Delete
    fun deleteRecommendationActivities(activities: List<RecommendationActivity>)
}
